ggml/src/ggml-backend.cpp

```diff
@@ -1997,3 +1999,199 @@ ggml_backend_buffer_t ggml_backend_cpu_buffer_from_ptr(void * ptr, size_t size)
     GGML_ASSERT((uintptr_t)ptr % TENSOR_ALIGNMENT == 0 && "buffer pointer must be aligned");
     return ggml_backend_buffer_init(ggml_backend_cpu_buffer_from_ptr_type(), ggml_backend_cpu_buffer_from_ptr_i, ptr, size);
 }
+
+static void ggml_backend_sched_splits_fdump_dot(FILE * fp, ggml_backend_sched_t sched, const struct ggml_cgraph * graph) {
+    std::set<void *> visited;
+
+    for (int i = 0; i < sched->n_splits; i++) {
+        struct ggml_backend_sched_split * split = &sched->splits[i];
+        ggml_backend_t split_backend = sched->backends[split->backend_id];
+
+        fprintf(fp, "  subgraph cluster_%d {"
+                    "    node [style=filled];"
+                    "    label = \"SPLIT %d : %s # %d inputs\";\n",
+                i, i, ggml_backend_name(split_backend), split->n_inputs);
+
+        for (int j = split->i_start; j < split->i_end; j++) {
+            struct ggml_tensor *node = graph->nodes[j];
+            fprintf(fp, "    \"%p\";\n", (void *)node);
+            for (int k = 0; k < GGML_MAX_SRC; k++) {
+                struct ggml_tensor *src = node->src[k];
+                if (   (nullptr == src)
+                    || (tensor_backend_id(src) != split->backend_id)
+                    || (visited.find(src) != visited.end())) {
+                    continue;
+                }
+
+                visited.insert(src);
+                fprintf(fp, "    \"%p\";\n", (void *)src);
+            }
+        }
+        fprintf(fp, "  }\n");
+    }
+}
+
+static uint32_t simple_hash(const unsigned char *str)
+{
+    uint32_t hash = 5381;
+    while (unsigned char c = *str++) hash = hash * 33 + c;
+    return hash;
+}
+
+static const char *ggml_color_of_backend(ggml_backend_sched_t sched, struct ggml_tensor * node) {
```

Comment:
Just passing by here, I have some suggestions for this function:
- Maybe returning `int32_t` here is enough (we only use 3 bytes for 3 channels)
- A more simple approach without using hash would be to use `ggml_backend_sched_get_n_backends` to get number of all backends
  - Use HSL color space, divide the hue channel into N values
  - Increase lightness to get the light color (with saturation set to full)
  - Then, convert HSL to RGB
  - Here is a demo that I made: https://gist.github.com/ngxson/1cb649624d69201be16311c2a3d54060

Reply:
Hmm, maybe graphviz support HSV color out of the box (which is good if true - I haven't tested): https://graphviz.org/docs/attr-types/color/

So with HSV, my code will be reduced to just one line:

```
printf("%f,1.0,0.8", idx / (total + 1));
```

Reply:
> Just passing by here, I have some suggestions for this function:
> 
> * Maybe returning `int32_t` here is enough (we only use 3 bytes for 3 channels)
> * A more simple approach without using hash would be to use `ggml_backend_sched_get_n_backends` to get number of all backends
>   

It outputs two colors. So `int32_t` is not enough.

Reply:
In that case, you can return `std::pair`. This file is already cpp and the function is an internal one, so I think it would be cleaner not to have `static` variable

Reply:
No. Hue is a `float`.

Reply:
or `void ggml_print_color_of_backend(FILE * fp, ggml_backend_sched_t sched, struct ggml_tensor * node)`

Reply:
Updated to use `std::string`.